boat_simulator/nodes/low_level_control/controller.py

```diff
@@ -0,0 +1,293 @@
+
+        Args:
+            `target_angle` (Scalar): New desired heading
+
+        """
+        self.desired_heading = target_angle
+        self.prev_control.clear()
+        self.time_series.clear()
+
+    def get_heading(self) -> None:
+        """Returns the most recent heading and current running time"""
+        return (self.prev_control[-1], self.time_series[-1])
+
+    @abstractmethod
+    def update_state(self, speed: Scalar) -> None:
+        """Updates the controller position based on target heading
+
+        Args:
+            `speed` (Scalar): Iterative speed in degrees / second
+
+        """
+
+    @abstractmethod
+    def compute_feedback_angle(self) -> None:
+        """Computes the corresponding feedback based on error and desired heading"""
+        pass
+
+
+class RudderController(ActuatorController):
+    """General Class for the Actuator Controller.
+
+    RudderController Extends: ActuatorController
+    """
+
+    def __innit__(
+        self,
+        current_heading: Scalar,
+        desired_heading: Scalar,
+        current_control_ang: Scalar,
+        time_step: Scalar,
+        kp: Scalar,
+        cp: Scalar,
+        max_angle_range=(45, -45),
+    ):
+        """Initializes the class attributes.
+
+        Args:
+            `current_heading` (Scalar): Current boat heading direction in degrees.
+            `desired_heading` (Scalar): Target boating heading direction in degrees.
+            `current_control_ang` (Scalar): Current control mechanism angle in degrees.
+            `time_step` (Scalar): Time per iteration given in degrees/second.
+            `kp` (Scalar): Proportional constant when calculating error.
+            `cp` (Scalar): Timeseries of error values computed over time.
+            `max_angle_range` (tuple): Tuple of minimum to maximum control angle range
+
+        """
+        super().__init__(current_heading, desired_heading, current_control_ang, time_step)
+
+        self.kp = kp
+        self.cp = cp
+        self.max_angle_range = max_angle_range  # passed in degrees
+        self.running_error = 0
+
+    def compute_error(
+        self, desired: Scalar, current: Scalar
+    ) -> Scalar:  # angle passed in as radians
+        """Computes the error between desired and current heading
+        implementation taken from: https://stackoverflow.com/a/2007279
+
+        Args:
+            `desired` (Scalar): New desired heading
+            `current` (Scalar): New current heading
+
+        Returns:
+            Scalar: The error between the given headings in radians.
+        """
+        current_bound = bound_to_180(current)
+        desired_bound = bound_to_180(desired)
+        desired_rad = np.deg2rad(desired)
+        current_rad = np.deg2rad(current)
+        error = atan2(sin(desired_rad - current_rad), cos(desired_rad - current_rad))
+
+        if current_bound == desired_bound:
+            error = 0
```

Comment:
Is this potentially reduntant to check twice?

Reply:
I think I thought it was originally, but this was from my code testing from the PID, and I think I was running into some random edge cases, but I'll recheck again.